### ISSUE TITLE
How to Convert YOLOv10 Model to TFLite with INT8 Quantization?

### ISSUE THREAD

**Issue by AhmedFkih** [THU-MIG/yolov10#176]
Hi everyone,

I’m working on a project that involves deploying a YOLOv10 model on a mobile/edge device. To improve inference speed and reduce the model size, I want to convert my YOLOv10 model to TensorFlow Lite (TFLite) with INT8 quantization.

-What is the best approach to convert the YOLOv10 model to TFLite with INT8 quantization?
-Are there any specific calibration techniques required for YOLOv10 during quantization?
-Can someone provide a code example or point me to resources for performing INT8 quantization on YOLOv10?

**Comment by senceryucel** [comment]
Hi @AhmedFkih @nenenkosi 

You can use the following code that I used for YOLOX more than a year ago. I can't remember the versions of the packages, but I think you can easily solve incompatibility issues.

Do not forget that you need to convert the YOLOv10 model from pt to onnx first ([check this](https://pytorch.org/tutorials/beginner/onnx/export_simple_model_to_onnx_tutorial.html)).

Please be aware that yo may need to change the parameters of the quantize method (quantize_static), since my requirements were probably way different than yours. You should also check for quantize_dynamic method which you can import from onnxruntime.quantization as well as quantize_static.

```
import numpy as np
from onnxruntime.quantization import quantize_static, CalibrationMethod, CalibrationDataReader, QuantType, QuantFormat

# loading the ONNX model
onnx_model_input_path = "YOLOv10.onnx"
onnx_model_output_path = "output.onnx"

# calibration dataset (dummy data for calibration)
class DummyDataReader(CalibrationDataReader):
    def __init__(self, num_samples):
        self.num_samples = num_samples
        self.current_sample = 0

    def get_next(self):
        if self.current_sample < self.num_samples:
            input_data = self.generate_random_input()
            self.current_sample += 1
            return {'images': input_data}
        else:
            return None

    def generate_random_input(self):
        input_data = np.random.uniform(-1, 1, size=input_shape).astype(np.float32)
        return input_data

num_calibration_samples = 100
input_shape = (1, 3, 640, 640)

calibration_data_reader = DummyDataReader(num_samples=num_calibration_samples)


# Quantize the model to int8
quantized_model = quantize_static(
    model_input=onnx_model_input_path,
    model_output=onnx_model_output_path,
    calibration_data_reader=calibration_data_reader,
    activation_type=QuantType.QInt8,
    weight_type=QuantType.QInt8,
    quant_format=QuantFormat.QDQ,
    per_channel=False,
    calibrate_method=CalibrationMethod.MinMax
)
```

**Comment by rabion1234** [comment]
hey @senceryucel 

i want my converted onnx model from yolov10 to tflite but your code will give onnx INT8 Quantization.

**Comment by enzoferrari1** [comment]
Hi! Was this solved? I'm facing the same issue

**Comment by bluesy7585** [comment]
use onnx2tf tool can do both tflite conversion and int8 post-training quantization.
- export onnx model
- use onnx2tf to create int8 tflite from exported onnx model.
`onnx2tf -i ./yolov10s.onnx -o ./best_saved_mdel -nuo --verbosity info -oiqt -ioqd uint8 -qt per-tensor -cind images [path to calibration npy file] "[[[[0, 0, 0]]]]" "[[[[1, 1, 1]]]]"`

ps. i have modified head.py for adding bbox coordination normalization. (x1y1x2x2 to 0~1) to make sure bbox and conf in the same value range. otherwise quantization will not work.  post-process also removed in my case.

**Comment by enzoferrari1** [comment]
I'll leave what i've done in case what bluesy7585 solution didn't work.

Use ultralytics method to export as onnx format
```
from ultralytics import YOLOv10
model = YOLOv10('/content/best-100-only-card.pt')
model.export(format='onnx')
```

Use [onnx2tflite](https://github.com/MPolaris/onnx2tflite) to convert to tflite format
```
!git clone https://github.com/MPolaris/onnx2tflite.git
%cd onnx2tflite
!pip install -r requirements.txt
```

Save the model
`
!python converter.py --weights "/content/best-100-only-card.onnx" --outpath "./" --formats "tflite"`

**Comment by kurreman** [comment]
Have you simply tried this?:

`from ultralytics import YOLO

# Load the YOLOv10 model
model = YOLO('yolov10n.pt')

# Export the model to TFLite format
model.export(format="tflite",int8=True,data="path/to/calibration_dataset")`

The calibration dataset is not always needed, default is a subset of COCO

**Comment by p-robak97** [comment]
@bluesy7585 Could you please share the code with the changes you made?

**Comment by bluesy7585** [comment]
> @bluesy7585 Could you please share the code with the changes you made?

in head.py L59, change condition to always export with normalization. 
```
        if self.export:# and self.format in ("tflite", "edgetpu"):
            # Precompute normalization factor to increase numerical stability
            # See https://github.com/ultralytics/ultralytics/issues/7371
```
and in v10Detect forward method, return without postprocess
(this is because I had error about Tile layer in postprocess during onnx2tf conversion, but I think its fixed in recent onnx2tf release )
```
           if not self.export:
                return {"one2many": one2many, "one2one": one2one}
            else:
                assert(self.max_det != -1)
                return one2one 
                #boxes, scores, labels = ops.v10postprocess(one2one.permute(0, 2, 1), self.max_det, self.nc)
                #return torch.cat([boxes, scores.unsqueeze(-1), labels.unsqueeze(-1).to(boxes.dtype)], dim=-1)
```

**Comment by Egorundel** [comment]
You can try my C++ code on quantization and TensorRT Engine creation.

https://github.com/Egorundel/int8_calibrator_cpp

**Comment by egSat** [comment]
Hi all

I have this error when trying to convert yolov10s to tflite with int8.
Command: 
onnx2tf -i /runs/train/weights/best.onnx -o /runs/train/weights/ -nuo --verbosity info -oiqt -ioqd uint8 -qt per-tensor -cind images /runs/train/weights/best_saved_model/tmp_tflite_int8_calibration_images.npy "[[[[0, 0, 0]]]]" "[[[[1, 1, 1]]]]


alueError: A KerasTensor cannot be used as input to a TensorFlow function. A KerasTensor is a symbolic placeholder for a shape and dtype, used when constructing Keras Functional models or Keras Functions. You can only use it as input to a Keras layer or a Keras operation (from the namespaces `keras.layers` and `keras.operations`). You are likely doing something like:

```
x = Input(...)
...
tf_fn(x)  # Invalid.
```

What you should do instead is wrap `tf_fn` in a layer:

```
class MyLayer(Layer):
    def call(self, x):
        return tf_fn(x)

x = MyLayer()(x)
```


Aynoue have encountered the same?

thanks!
